galsim/table.py

```diff
@@ -645,76 +666,110 @@ def _call_constant(self, x, y):
         else:
             x = np.array(x, dtype=float, copy=False)
             y = np.array(y, dtype=float, copy=False)
-            f = np.empty_like(x, dtype=float)
-            f.fill(self.constant)
-            good = ((x >= self.x[0]) & (x <= self.x[-1]) &
-                    (y >= self.y[0]) & (y <= self.y[-1]))
-            xx = np.ascontiguousarray(x[good].ravel(), dtype=float)
-            yy = np.ascontiguousarray(y[good].ravel(), dtype=float)
-            tmp = np.empty_like(xx, dtype=float)
-            self._tab.interpMany(xx.ctypes.data, yy.ctypes.data, tmp.ctypes.data, len(xx))
-            f[good] = tmp
-            return f
-
-    def _call_raise(self, x, y):
+            if grid:
+                f = np.empty((len(x), len(y)), dtype=float)
+                # Fill in interpolated values first, then go back and fill in
+                # constants
+                xx = np.ascontiguousarray(x.ravel(), dtype=float)
+                yy = np.ascontiguousarray(y.ravel(), dtype=float)
+                self._tab.interpGrid(xx.ctypes.data, yy.ctypes.data, f.ctypes.data, len(xx), len(yy))
+                badx = (x < self.x[0]) | (x > self.x[-1])
+                bady = (y < self.y[0]) | (y > self.y[-1])
+                f[badx,:] = self.constant
+                f[:, bady] = self.constant
+                return f
+            else:
+                # Start with constant array, then interpolate good positions
+                f = np.empty_like(x, dtype=float)
+                f.fill(self.constant)
+                good = ((x >= self.x[0]) & (x <= self.x[-1]) &
+                        (y >= self.y[0]) & (y <= self.y[-1]))
+                xx = np.ascontiguousarray(x[good].ravel(), dtype=float)
+                yy = np.ascontiguousarray(y[good].ravel(), dtype=float)
+                tmp = np.empty_like(xx, dtype=float)
+                self._tab.interpMany(xx.ctypes.data, yy.ctypes.data, tmp.ctypes.data, len(xx))
+                f[good] = tmp
+                return f
+
+    def _call_raise(self, x, y, grid=False):
         if not self._inbounds(x, y):
             raise GalSimBoundsError("Extrapolating beyond input range.",
                                     PositionD(x,y), self._bounds)
```

Comment:
I think I missed this before, but won't these not work right when x,y are arrays?  Similarly other places that do `PositionD(x,y)`.

Reply:
Yep, good find.  I hacked together a solution to find a representative out-of-bounds point and raise an exception using that.